src/connector/src/sink/batching_log_sink.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_trait::async_trait;
+
+use crate::sink::file_sink::opendal_sink::OpenDalSinkWriter;
+use crate::sink::log_store::{LogStoreReadItem, TruncateOffset};
+use crate::sink::writer::SinkWriter;
+use crate::sink::{LogSinker, Result, SinkLogReader};
+
+/// `BatchingLogSinkerOf` is used for a commit-decoupled sink that supports cross-barrier batching.
+/// Currently, it is only used for file sinks, so it contains an `OpenDalSinkWriter`.
+/// When the file sink writer completes writing a file, it truncates the corresponding `chunk_id` in the log store.
+pub struct BatchingLogSinkerOf {
+    writer: OpenDalSinkWriter,
+}
+
+impl BatchingLogSinkerOf {
+    /// Create a log sinker with a file sink writer.
+    pub fn new(writer: OpenDalSinkWriter) -> Self {
+        BatchingLogSinkerOf { writer }
+    }
+}
+
+#[async_trait]
+impl LogSinker for BatchingLogSinkerOf {
+    async fn consume_log_and_sink(self, log_reader: &mut impl SinkLogReader) -> Result<!> {
+        let mut sink_writer = self.writer;
+        #[derive(Debug)]
+        enum LogConsumerState {
+            /// Mark that the log consumer is not initialized yet
+            Uninitialized,
+
+            /// Mark that a new epoch has begun.
+            EpochBegun { curr_epoch: u64 },
+
+            /// Mark that the consumer has just received a barrier
+            BarrierReceived { prev_epoch: u64 },
+        }
+
+        let mut state = LogConsumerState::Uninitialized;
+        loop {
+            let (epoch, item): (u64, LogStoreReadItem) = log_reader.next_item().await?;
+            if let LogStoreReadItem::UpdateVnodeBitmap(_) = &item {
+                match &state {
+                    LogConsumerState::BarrierReceived { .. } => {}
```

Comment:
I think we need to force to finish the batch (i.e. commit) here. Otherwise, there can be data loss because actor can be dropped and rebuilt during scaling.

Reply:
Got it.
> we need to force to finish the batch
In this case, should we 
1. force commit the written data regardless of whether it trigger batching strategy
2. only commit those triggered batching chunk

I think we need to do 1, right?